### PR TITLE
Request workshop rework for FAC17

### PR DESCRIPTION
**Pull request by starsuit**
FAC16 (and I think 15 too) struggled with this as it was pretty confusing. 

See #7  and [this master reference issue](https://github.com/foundersandcoders/master-reference/issues/978) for arguments about removing dependency/reference to the request module.

I've rewritten this workshop so that it doesn't reference the request module at all, just asks for a function which takes a url and a callback, and makes an http.get to that url.

I've also written some tests with tape/tap-spec/nodemon which should check the function is working correctly (they start off failing, students should try and make them pass). There's a commented-out test for the bonus https part.

There's also a solution folder and file, which could replace the solutions branch on this repo.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,35 +1,46 @@
 # Morning challenge - Build the Request module
-This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and it's methods. Now that you are in the backend we would like you to use the Request module which is [available on npm](https://www.npmjs.com/package/request). This workshop is intended to familiarise you with how this module works by asking you to build a simple version of it.
+This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and methods. Now that you are in the backend we would like you to learn how to make similar request using the Node core [http module](https://nodejs.org/api/http.html). This workshop is intended to familiarise you with how backend requests are made, and practice using callbacks.
```
Comment (oliverjam) [inline README.md]:
```suggestion
This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and methods. Now that you are in the backend we would like you to learn how to make similar requests using the Node core [http module](https://nodejs.org/api/http.html). This workshop is intended to familiarise you with how backend requests are made, and practice using callbacks.
```

File: README.md [hunk 2]
```diff
@@ -1,35 +1,46 @@
 # Morning challenge - Build the Request module
```
Comment (oliverjam) [inline README.md]:
```suggestion
# Morning challenge - Build a request module
```

File: README.md [hunk 3]
```diff
@@ -1,35 +1,46 @@
...
-This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and it's methods. Now that you are in the backend we would like you to use the Request module which is [available on npm](https://www.npmjs.com/package/request). This workshop is intended to familiarise you with how this module works by asking you to build a simple version of it.
+This week in your project you will be making API calls from the backend. In the past few weeks you have been making API calls from the frontend with the XMLHttpRequest object and methods. Now that you are in the backend we would like you to learn how to make similar request using the Node core [http module](https://nodejs.org/api/http.html). This workshop is intended to familiarise you with how backend requests are made, and practice using callbacks.
 
 ### Time Required: 1 hour
```
Comment (oliverjam) [inline README.md]: Not sure this section is needed. Students should know how a morning challenge goes by Week 5

File: README.md [hunk 4]
```diff
@@ -1,35 +1,46 @@
...
   - 40 mins to complete the challenge in pairs
   - 10 mins to regroup and talk over the solution
 
 ### Learning outcome
-To become familiar with the Request module and the Node core [http module](https://nodejs.org/api/http.html).
+To become familiar with making API calls from the backend using the Node core [http module](https://nodejs.org/api/http.html).
 
-### What is the Node Request module?
-Request is a very popular module that helps to make HTTP requests as simple as possible and it is built around the Node HTTP core module. [Here](http://stackabuse.com/the-node-js-request-module/) is a brilliant article on the Request module.
+### When would I need to make an API call from the backend?
```
Comment (oliverjam) [inline README.md:7]: Maybe the examples should highlight why students might want to make an API call _from the backend_. E.g bypassing CORS or hiding API keys

File: README.md [hunk 5]
```diff
@@ -1,35 +1,46 @@
...
+### The challenge
 
-### Your challenge
-Create a function called **myRequest** that has similar set of features to the Node Request module.
+Create a function called `myRequest` which makes a request to an API url using an http GET request.
 
-The npm Request module simply creates a wrapper around the Node core http module. Your task it to create your own request function `myRequest` that performs in the same way as the Request npm module, which is used in the app.js file in this repo.
+It should take two arguments: the url to make the request to, and a callback function.
+
+The callback function will have two arguments:
```
Comment (oliverjam) [inline README.md]:
```suggestion
The callback function should take two arguments:
```

File: README.md [hunk 6]
```diff
@@ -1,35 +1,46 @@
...
+
+  1. error (String: if an error occurred)
```
Comment (oliverjam) [inline README.md]:
```suggestion
  1. error (null if no error occurred; an error object if an error occurred)
```

File: README.md [hunk 7]
```diff
@@ -1,35 +1,46 @@
...
 
 **Hint**: relate to your knowledge of streams and sending data in chunks.
 
 To complete this challenge you will need to use the [`http.get`](https://nodejs.org/api/http.html#http_http_get_options_callback) method of the node core `http` module.
 
 ### Start here!
 1) Clone this repo & run `npm install`
-2) Run the app.js file in your command line with the command `node app.js`. This is the response from the request module that you're trying to recreate with your myRequest function. There is no need to change the helper function code.
-3) Open app.js and follow the instructions to create your own request function. 
+2) Run `npm test` in your command line to start the test watcher. The tests should fail as your function doesn't exist yet
```
Comment (oliverjam) [inline README.md]:
```suggestion
2) Run `npm test` in your command line to start the test watcher. The tests should fail as your function doesn't do anything yet
```

File: README.md [hunk 8]
```diff
@@ -1,35 +1,46 @@
...
+There are two test sets in `app.test.js` (the second is commented out). The first test will run three assertions: 
+1) Checks there are no errors
+2) Checks that the API response passed to the callback has a status code of 200
+3) Checks that the API response passed to the callback has the correct json
+
+The second, commented-out test will check your function works with https requests (bonus round, see below...). Uncomment it when you're ready to test in the bonus round!
 
 ### Bonus round
-Can you alter your myRequest function to check which protocol the url is using (http or https) and make the suitable request. You will need to require in the `https` node core module. You can test this by using the `https://jsonplaceholder.typicode.com/users/1` url.
+You can alter your myRequest function to check which protocol the url is using (http or https) and make the suitable request. You will need to require in the `https` Node core module. You can test this by uncommenting the second test in `app.test.js`.
```
Comment (oliverjam) [inline README.md]: Maybe add an intro to this asking them to try making a request to the Pokeapi (or anything that uses https) so they can see it fail

Comment (oliverjam) [inline README.md]:
```suggestion
You can alter your `myRequest` function to check which protocol the url is using (http or https) and use the appropriate Node core module. You will need to require in `https`. You can test this by uncommenting the second test in `app.test.js`.
```

File: README.md [hunk 9]
```diff
@@ -1,35 +1,46 @@
...
 To complete this challenge you will need to use the [`http.get`](https://nodejs.org/api/http.html#http_http_get_options_callback) method of the node core `http` module.
 
 ### Start here!
 1) Clone this repo & run `npm install`
-2) Run the app.js file in your command line with the command `node app.js`. This is the response from the request module that you're trying to recreate with your myRequest function. There is no need to change the helper function code.
-3) Open app.js and follow the instructions to create your own request function. 
+2) Run `npm test` in your command line to start the test watcher. The tests should fail as your function doesn't exist yet
+3) Open `app.js` and write your function!
+
+### Tests
```
Comment (oliverjam) [inline README.md]: Not sure if this section will be helpful or just confusing. They can always read the tests if the instructions above aren't clear enough

**Review by oliverjam — CHANGES_REQUESTED** [review]

**Comment by oliverjam** [comment]
It might be a bit too much for them at this point, but should we encourage them to use `http.request` so they have to set the method too? Then they could use their module for `POST`s etc too

Maybe that should be a stretch goal. Writing the body to the stream is a bit complicated

File: README.md [hunk 10]
```diff
@@ -1,35 +1,46 @@
...
-Create a function called **myRequest** that has similar set of features to the Node Request module.
+Create a function called `myRequest` which makes a request to an API url using an http GET request.
 
-The npm Request module simply creates a wrapper around the Node core http module. Your task it to create your own request function `myRequest` that performs in the same way as the Request npm module, which is used in the app.js file in this repo.
+It should take two arguments: the url to make the request to, and a callback function.
+
+The callback function should take two arguments:
+
+  1. error (null if no error occurred; an error object if an error occurred)
+  2. response (Object: includes the body & statusCode of the request)
```
Comment (bantek89) [inline README.md:21]: It would be helpful to add a hint that you should be constructing your own response object here

Comment (oliverjam) [reply README.md:21]: They don't _have_ to make a new object—the Node `response` object already has the `statusCode` so they could just add the `body` to it and use that

Comment (starsuit) [reply README.md:21]: Probably deserves some kind of note/hint, it's something we both got stuck on trying out the new workshop

Comment (oliverjam) [reply README.md:21]: I guess this is why the old workshop had the callback take 3 arguments: `(error, response, body)`

A note/hint is probably a good idea

File: README.md (hunk 3 above)
Comment (starsuit) [reply README.md]: [Removed](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/922e75e8d598eea525c7484ac5217f0563fa7ca9)

File: README.md (hunk 4 above)
Comment (starsuit) [reply README.md:7]: [Updated](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/d3c8b3c6b7bc888a43eb793c02c47529fea5d22e)

File: README.md (hunk 8 above)
Comment (starsuit) [reply README.md]: I've added a more detailed intro, how's [this](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/861b2debb769660884c075ecf50a3078c1daa126)?

File: solution/solution.js [hunk 11]
```diff
@@ -2,8 +2,8 @@ const http = require("http");
 const https = require("https");
 
 const myRequest = (url, cb) => {
-  const urlObj = new URL(url);
-  const protocol = urlObj.protocol === "https:" ? https : http;
+  // the line below is extra - the solution to the bonus section
+  const protocol = url.includes("https") ? https : http;
```
Comment (oliverjam) [inline solution/solution.js]: Maybe it would be clearer to have a `bonus.solution.js`?

File: app.js [hunk 12]
```diff
@@ -1,29 +1,10 @@
-"use strict";
```
Comment (astroash) [inline app.js]: Can we keep `"use strict";` in?

File: app.js [hunk 13]
```diff
@@ -1,29 +1,10 @@
...
-
-// This is the npm Request module that you are trying to emulated with `myRequest` below:
-const request = require('request');
-
 const myRequest = () => {
   /*
   create your own request module here.
-  It should take a url to make a http GET request, and a callback function with three arguments;
+  It should take a url to make a https GET request, and a callback function with two arguments;
```
Comment (astroash) [inline app.js]: should "https" be "http" here?

**Review by astroash — CHANGES_REQUESTED** [review]
This looks great! Agree with @oliverjam that the solutions file will be a little clearer if we have a separate commented out solution for the bonus section

File: app.js (hunk 13 above)
Comment (starsuit) [reply app.js]: fixed, thanks!

File: app.js (hunk 12 above)
Comment (starsuit) [reply app.js]: I've re-added it

File: solution/solution.js (hunk 11 above)
Comment (starsuit) [reply solution/solution.js]: I've moved the bonus solution into a new file - also added tests for the solution files

File: README.md (hunk 9 above)
Comment (starsuit) [reply README.md]: When @bantek89 tried this out for practice I think she found the tests useful? We could see how FAC17 find it and remove if it is confusing?

File: README.md (hunk 10 above)
Comment (starsuit) [reply README.md:21]: I've added a [hint](https://github.com/foundersandcoders/mc-request-module-workshop/pull/8/commits/1774c5c438a3338130fcd2efa4634071aee071e1)

**Comment by starsuit** [comment]
> It might be a bit too much for them at this point, but should we encourage them to use `http.request` so they have to set the method too? Then they could use their module for `POST`s etc too
> 
> Maybe that should be a stretch goal. Writing the body to the stream is a bit complicated

@oliverjam I think this might be too much at once. Mentors could mention it if they find that some students have finished the workshop early.

File: README.md (hunk 9 above)
Comment (oliverjam) [reply README.md]: I agree that the tests are useful, I just don't know if a section explaining what the tests do adds much. If the tests themselves are confusing then maybe they should have comments next to them in the code?

Comment (starsuit) [reply README.md]: OK I've removed that section of the readme

**Review by oliverjam — APPROVED** [review]

**Review by astroash — APPROVED** [review]

**Pull request merged**